lib/Transforms/LayoutOptimization/LayoutOptimization.cpp

```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
+static Cost computeKernelCostFromDAG(KernelName kernel,
+                                     ArrayRef<int64_t> shape) {
+  using NodeTy = kernel::ArithmeticDagNode<kernel::SymbolicValue>;
+  std::shared_ptr<NodeTy> kernelDag;
+
+  switch (kernel) {
+    case KernelName::Trivial:
+      return 0;
+
+    case KernelName::MatvecDiagonal: {
+      if (shape.size() < 2) return 0;
+      // Create symbolic matrix (rows x cols) and vector (cols)
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      kernelDag =
+          kernel::implementMatvec(kernel, symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::MatvecNaive: {
+      if (shape.size() < 2) return 0;
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      // MatvecNaive not yet implemented in KernelImplementation.h
+      // For now, use diagonal as approximation
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::VecmatDiagonal: {
+      if (shape.size() < 2) return 0;
+      // VecmatDiagonal: vector (rows) * matrix (rows x cols)
+      // Approximation: similar structure to MatvecDiagonal but transposed
+      kernel::SymbolicValue symbolicVector({shape[0]});
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      // Use same cost as matvec for now (same number of diagonals)
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
+
+    case KernelName::MatmulDiagonal: {
+      if (shape.size() < 2) return 0;
+      // MatmulDiagonal typically implemented via multiple matvecs
+      // Conservative estimate: cost scales with both dimensions
+      return shape[0] + shape[1];
+    }
+
+    default:
+      return 0;
+  }
+
+  if (!kernelDag) {
+    return 0;
+  }
+
+  // Count rotations in the symbolic DAG with CSE deduplication
+  KernelRotationCountVisitor counter;
+  return counter.process(kernelDag);
+}
+
 static Cost costOfKernelChange(Operation* op, KernelName oldKernel,
                                const HoistResult& hoistResult) {
-  // TODO(#2316): add the cost of a kernel change.
-  return 0;
+  KernelName newKernel = hoistResult.newKernel;
+
+  // No cost if kernel isn't changing
+  if (oldKernel == newKernel) {
+    return 0;
+  }
+
+  // Trivial kernels are free (elementwise operations)
+  if (newKernel == KernelName::Trivial) {
+    return 0;
+  }
+
+  // Extract operation dimensions
+  auto shape = getOperationShape(op);
+  if (!shape.has_value()) {
+    LLVM_DEBUG(llvm::dbgs() << "Could not extract shape for kernel cost\n");
+    return 0;  // Conservative: unknown ops cost nothing
```

Comment:
Perhaps a large cost would be more conservative option here: if you use zero then the "unknown op" would be more likely to be chosen by the optimizer. Also better than LLVM_DEBUG is to use `op.emitError() << "..."` since this should alert the person running the  compiler that something is wrong.